libsolutil/CommonData.h

```diff
@@ -267,17 +267,21 @@ template<
 	typename MapType,
 	typename KeyType,
 	typename ValueType = std::decay_t<decltype(std::declval<MapType>().find(std::declval<KeyType>())->second)> const&,
-	typename AllowCopyType = void*
+	typename AllowCopyType = std::conditional_t<std::is_pod_v<ValueType> || std::is_pointer_v<ValueType>, detail::allow_copy, void*>
 >
-decltype(auto) valueOrDefault(MapType&& _map, KeyType const& _key, ValueType&& _defaultValue = {}, AllowCopyType = nullptr)
+decltype(auto) valueOrDefault(
+	MapType&& _map,
+	KeyType const& _key,
+	ValueType&& _defaultValue = {},
+	AllowCopyType = {})
 {
 	auto it = _map.find(_key);
 	static_assert(
 		std::is_same_v<AllowCopyType, detail::allow_copy> ||
-		std::is_reference_v<decltype((it == _map.end()) ? _defaultValue : it->second)>,
+		std::is_reference_v<decltype((it == _map.end()) ? std::forward<ValueType>(_defaultValue) : it->second)>,
 		"valueOrDefault does not allow copies by default. Pass allow_copy as additional argument, if you want to allow copies."
 	);
-	return (it == _map.end()) ? _defaultValue : it->second;
+	return (it == _map.end()) ? std::forward<ValueType>(_defaultValue) : it->second;
```

Comment:
The forwarding was definitely missing - without it the ``_defaultValue`` always counts as reference and ``is_reference_v`` returns ``true`` (if no conversion needs to happen) and it breaks for temporaries...

Reply:
What does `std::forward<ValueType>(_defaultValue)` do, if `_defaultValue` is a temporary? I am guessing it returns a temporary of `ValueType`?

Reply:
Yeah - the result is that ``std::forward<ValueType>(_defaultValue)`` will not be an lvalue reference, so the entire expression ``(it == _map.end()) ? std::forward<ValueType>(_defaultValue) : it->second`` will have to copy the value in the map, if it's there...